src/table.js

```diff
@@ -488,30 +489,44 @@ Table.prototype.createReadStream = function(options) {
   if (options.limit) {
     reqOpts.rowsLimit = options.limit;
   }
-
+  const chunkFormatter = new ChunkFormatter();
   var stream = pumpify.obj([
     this.requestStream(grpcOpts, reqOpts),
-    through.obj(function(data, enc, next) {
-      var throughStream = this;
-      var rows = Row.formatChunks_(data.chunks, {
-        decode: options.decode,
-      });
-
-      rows.forEach(function(rowData) {
-        if (stream._ended) {
-          return;
+    through.obj(
+      function(data, enc, next) {
+        var throughStream = this;
+        try {
+          chunkFormatter.formatChunks(
+            data.chunks,
+            {
+              decode: options.decode,
+            },
+            (err, rowData) => {
+              if (stream._ended) {
+                return false;
+              } else {
+                var row = self.row(rowData.key);
+                row.data = rowData.data;
+                throughStream.push(row);
+                return true;
+              }
+            }
+          );
+        } catch (err) {
+          return next(err);
         }
-
-        var row = self.row(rowData.key);
-
-        row.data = rowData.data;
-        throughStream.push(row);
-      });
-
-      next();
-    }),
+        next();
+      },
+      function(callback) {
+        try {
```

Comment:
It would be nice to remove the try/catch here as well, something more like:

```js
through.obj(
  function(data, enc, next) {
    // ...
    next();
  },
  chunkFormatter.onStreamEnd
);
```

`onStreamEnd` should accept a callback, and return an error if it ended prematurely.